infinispan-embedded/src/main/java/org/hibernate/ogm/datastore/infinispan/impl/InfinispanEmbeddedStoredProceduresDialect.java

```diff
@@ -0,0 +1,152 @@
+		if ( positionalParameters != null && positionalParameters.size() > 0 ) {
+			throw log.dialectDoesNotSupportPositionalParametersForStoredProcedures( getClass() );
+		}
+	}
+
+	private Object execute(String storedProcedureName, EmbeddedCacheManager embeddedCacheManager, Callable<?> callable) {
+		AtomicReference<Object> ref = new AtomicReference<>();
+		try {
+			return embeddedCacheManager.executor()
+					.submitConsumer( ecm -> execute( callable ), ( a, r, e ) -> {
+						if ( e != null ) {
+							if ( e instanceof ExceptionWrapper ) {
+								throw log.cannotExecuteStoredProcedure( storedProcedureName, e.getCause() );
+							}
+							throw log.cannotExecuteStoredProcedure( storedProcedureName, e );
+						}
+						ref.compareAndSet( null, r );
+					} )
+					.thenCompose( v -> CompletableFuture.supplyAsync( ref::get ) )
+					.get();
+		}
+		catch ( Exception e ) {
+			throw log.cannotExecuteStoredProcedure( storedProcedureName, e );
+		}
+	}
+
+	private static Callable<?> instantiate(String storedProcedureName, String className, ClassLoaderService classLoaderService ) {
+		try {
+			Class<?> clazz = classLoaderService.classForName( className );
+			return (Callable<?>) clazz.newInstance();
+		}
+		catch (Exception e) {
+			throw log.cannotInstantiateStoredProcedure( storedProcedureName, className, e );
+		}
+	}
+
+	private static void setParams( String storedProcedureName, ProcedureQueryParameters queryParameters, Callable<?> callable ) {
+			Map<String, Object> params = queryParameters.getNamedParameters();
+		for ( Map.Entry<String, Object> entry : params.entrySet() ) {
+			try {
+				ReflectionHelper.setField( callable, entry.getKey(), entry.getValue() );
+			}
+			catch (Exception e) {
+				throw log.cannotSetStoredProcedureParameter( storedProcedureName, entry.getKey(), entry.getValue(), e );
+			}
+		}
+	}
+
+	private static Object execute(Callable<?> callable) {
+		try {
+			return callable.call();
+		}
+		catch (Exception e) {
+			throw new ExceptionWrapper( e );
+		}
+	}
+
+	private static List<Tuple> extractTuples(String storedProcedureName, Object retvalObj) {
+		if ( retvalObj instanceof Iterable ) {
+			Iterable<?> it = (Iterable) retvalObj;
+			return StreamSupport.stream( it.spliterator(), false )
+					.map( r -> extractTuple( storedProcedureName, r ) )
+					.collect( Collectors.toList() );
+		}
+		Tuple tuple = new Tuple();
+		tuple.put( "result", retvalObj );
+		return Collections.singletonList( tuple );
+	}
+
+	private static Tuple extractTuple(String storedProcedureName, Object obj) {
+		try {
+			Tuple tuple = new Tuple();
+			Map<String, Object> introspect = ReflectionHelper.introspect( obj );
+			introspect.forEach( tuple::put );
+			return tuple;
+		}
+		catch (IntrospectionException | InvocationTargetException | IllegalAccessException e) {
+			throw log.cannotExtractStoredProcedureResultSet( storedProcedureName, obj, e );
+		}
+	}
+
+	private static class ExceptionWrapper extends RuntimeException {
```

Comment:
Why do you need an ExceptionWrapper? Can't you throw an HibernateException?

Reply:
Thank you, we can. I forgot cleanup this workaround from my previous implementation